azure/services/inboundnatrules/inboundnatrules.go

```diff
@@ -19,133 +19,98 @@ package inboundnatrules
 import (
 	"context"
 
-	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
-	"github.com/Azure/go-autorest/autorest/to"
 	"github.com/go-logr/logr"
 	"github.com/pkg/errors"
 
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-02-01/network"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
-	"sigs.k8s.io/cluster-api-provider-azure/azure/services/loadbalancers"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/services/async"
+	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
+const serviceName = "inboundnatrules"
+
 // InboundNatScope defines the scope interface for an inbound NAT service.
 type InboundNatScope interface {
 	logr.Logger
 	azure.ClusterDescriber
-	InboundNatSpecs() []azure.InboundNatSpec
+	// azure.NetworkDescriber
+	azure.AsyncStatusUpdater
+	APIServerLBName() string
+	InboundNatSpecs([]network.InboundNatRule) []azure.ResourceSpecGetter
 }
 
 // Service provides operations on Azure resources.
 type Service struct {
 	Scope InboundNatScope
 	client
-	loadBalancersClient loadbalancers.Client
 }
 
 // New creates a new service.
 func New(scope InboundNatScope) *Service {
 	return &Service{
-		Scope:               scope,
-		client:              newClient(scope),
-		loadBalancersClient: loadbalancers.NewClient(scope),
+		Scope:  scope,
+		client: newClient(scope),
 	}
 }
 
 // Reconcile gets/creates/updates an inbound NAT rule.
 func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "inboundnatrules.Service.Reconcile")
-	defer done()
-
-	for _, inboundNatSpec := range s.Scope.InboundNatSpecs() {
-		s.Scope.V(2).Info("creating inbound NAT rule", "NAT rule", inboundNatSpec.Name)
-
-		lb, err := s.loadBalancersClient.Get(ctx, s.Scope.ResourceGroup(), inboundNatSpec.LoadBalancerName)
-		if err != nil {
-			return errors.Wrapf(err, "failed to get Load Balancer %s", inboundNatSpec.LoadBalancerName)
-		}
+	ctx, span := tele.Tracer().Start(ctx, "inboundnatrules.Service.Reconcile")
```

Comment:
let's keep using tele.StartSpanWithLogger everywhere

it allows us to get a custom logger with the correlation ID (see #1806)